### PR TITLE
Seeker interface implementation

### PR DESCRIPTION
**Pull request by Xmister**
I needed seeking capabilities for a project, so I implemented it.
I was trying to make it in the least "intrusive" way I could think of.

I'm mostly using it with http served contents, and it works well with `Range` requests.

Sadly, the new tests could only run if I had changed the include to my repo. That will need to be changed back after merge.

### CODE REVIEW

File: reader.go
Comment (klauspost) [inline reader.go]:
```suggestion
	SeekStart   = io.SeekStart // seek relative to the origin of the file
```
etc.

File: reader.go
Comment (klauspost) [inline reader.go]: Let's make it a bit shorter:
```suggestion
type ReadSeekCloser interface {
```
(and adjust respective names)

File: reader.go
Comment (klauspost) [inline reader.go]: `// NewReadSeeker creates a readahead Reader that also supports seeking.`

File: reader.go
Comment (klauspost) [inline reader.go]: Do we actually need to store this separately?

Couldn't we assert the interface on the `reader` when a seek is requested?

File: reader.go
Comment (klauspost) [inline reader.go]: Return early:
```
	if a.seeker == nil {
		return 0, fmt.Errorf("seeker not found")
	}
```

File: reader_test.go
Comment (klauspost) [inline reader_test.go]: Fix import

File: reader.go
Comment (klauspost) [inline reader.go]: This seems to me like it should be a `for` loop than ensures that all buffers are drained. Right now, as far as I can tell it will only add one block.

File: reader_test.go
Comment (klauspost) [inline reader_test.go]: Maybe make a bigger test buffer, to reads will be split into multiple reads.

Maybe a loop that tests various sizes and combination.

**Review by klauspost — COMMENTED** [review]
Thanks for the contribution! I have added some notes.

File: reader.go
Comment (Xmister) [reply reader.go]: The reason I defined them here is because `io.SeekStart`, etc. are not defined in Go1.6.

File: reader.go
Comment (Xmister) [reply reader.go]: I believe Go doesn't let you to cast an interface, so I don't see how we could force a `Reader` to be a `Seeker` also. I guess that"s the reason we also have a separate `closer` too.

File: reader_test.go
Comment (Xmister) [reply reader_test.go]: I think fixing this would require the `_test` file to be in the same readahead package. Are you ok with that?

File: reader.go
Comment (Xmister) [reply reader.go]: Well there is  a `for` loop right under this, which drains the ready channel, I might miss something, but I believe that's the only one that needs to be drained and taken into consideration.

File: reader.go
Comment (klauspost) [reply reader.go]: Ah, feel free to bump the travis script to Go 1.9, 1.10 and 1.11

File: reader.go
Comment (klauspost) [reply reader.go]: You cast with a check like this. `seeker, ok := in.(io.ReadSeeker)`  `if ok {` - we have a seeker.

In theory we don't even need special constructors for seekers.

**Comment by klauspost** [comment]
In general this could be made transparent to the caller.

Also, I don't like that we support the `io.Seeker` on the reader, but when called on a regular reader it will fail. Typically a programmer would think that if your returned value supports an interface, it will work.

A workaround could be:

```Go
type seekable struct {
   reader
}

func (*seekable) Seek(offset int64, whence int) (res int64, err error) {
...
}

...
// If a reader supporting the io.Seeker is given,
// the returned reader will also support it.
func NewReader(rd io.Reader) io.ReadCloser {
....
	if _, ok := rd.(io.Seeker); ok {
		return &seekable{reader: ret}
	}
	return ret
}
```

The main downside is that the caller would have to check for the interface, but making the promise above shouldn't be an issue.

It keeps the interface rather simple and doesn't overpromise.

File: reader_test.go
Comment (klauspost) [reply reader_test.go]: As a general tip, you can do the modifications in the original gopath position, but add your own repo as a secondary remote.

That way you can push to your own repo, while still retaining the original gopath. 

Obviously I cannot merge it whis way :)

**Comment by Xmister** [comment]
You have some very useful points, thank you.

For your last comment though: The Seeker interface is only available when using the Seeker aware constructors, so the error in the sanity check shouldn't really happen anytime.
But I think I'm going to use your idea to reduce duplicated code.

File: reader_test.go
Comment (Xmister) [reply reader_test.go]: @klauspost What do you think about the current approach?

**Review by klauspost — APPROVED** [review]
Thanks! I will merge once test complete!

**Pull request merged**
